src/datepicker/stateful-container.js

```diff
@@ -16,31 +16,35 @@ import type {
   StateReducerT,
 } from './types.js';
 
-const defaultStateReducer: StateReducerT = (type, nextState) => nextState;
+type InputProps<T> = CalendarPropsT<T> | DatepickerPropsT<T>;
 
-type PropsT = StatefulContainerPropsT<CalendarPropsT | DatepickerPropsT>;
-class StatefulContainer extends React.Component<PropsT, ContainerStateT> {
-  static defaultProps = {
+type PropsT<T> = StatefulContainerPropsT<InputProps<T>, T>;
+
+class StatefulContainer<T = Date> extends React.Component<
+  PropsT<T>,
+  ContainerStateT<T>,
+> {
+  static defaultProps: {stateReducer: StateReducerT<T>} = {
     initialState: {},
-    stateReducer: defaultStateReducer,
+    stateReducer: (type, nextState) => nextState,
     onChange: () => {},
   };
 
-  constructor(props: PropsT) {
+  constructor(props: PropsT<T>) {
     super(props);
-    const value = props.range ? [] : (null: ?Date);
+    const value = props.range ? [] : (null: ?T);
     this.state = {value, ...props.initialState};
   }
 
-  onChange = (data: {date: ?Date | Array<Date>}) => {
+  onChange: ({date: ?T | Array<T>}) => mixed = data => {
```

Comment:
There seems to be some buggy behavior in flow(noted [here](https://github.com/facebook/flow/issues/5193) and [here](https://github.com/facebook/flow/issues/4743)) related to the way that generic types flow through class methods. 

The only consistent solution is to switch to the kind of type declaration shown above.  The problem is, because the behavior is buggy, there doesn't actually seem to be a clear pattern as to which types of functions are a problem and which type of functions aren't.  I just ended up having to swap over every function that flow was complaining about, even if it didn't really make sense why.  It works as solution, it can just be a little confusing to figure out why one type signature had to change while another similar one didn't.